### PR TITLE
Consolidate gulp watch and gulp build

### LINKED ISSUE #12747: The entry point for locally served files is incorrectly presumed when consumed on a nonlocal server
## What's the issue?
The entry point for locally served files is incorrectly presumed (in `extension-location.js`) via the location object when consuming them on a nonlocal server. 

## How do we reproduce the issue?
- gulp
  - no --host config
- run secondary server 
- consume amp extension
  - Note: issue may only be visible with extensions that use web workers
  - In my scenario, I used `amp-bind`. I followed from the `AmpWorker` constructor to  `calculateEntryPointScriptUrl()`. I noticed it generated `ww.js` instead of `ww.max.js` with `serve=default` and that it was trying to load from my second server instead of localhost.

## What browsers are affected?
- all browsers

## Which AMP version is affected?
- version: 1516393161691

------
#### Potential fix
Add in check (probably in `calculateScriptBaseUrl()`) for whether or not a host config has been changed. If it has, proceed as it is currently, otherwise use `localhost`. Probably also check if port config has been changed.

### LINKED ISSUE #12852: isLocalDev applied 2 minutes after serving is confusing
The way `localDev` is currently set is at the end of the build during the write to dist/amp.js: 
```
[14:12:05] Enabled local development mode in dist/amp.js
[14:12:05] Wrote prod AMP config to dist/amp.js
```
This happens about 2 minutes after we start serving files locally, which results in extension developer confusion because during those 2 minutes, any extensions loading other extensions will try to load them from our CDN instead of localhost. 

Suggesting that we either print something to this effect in logs right before serving to localhost (please wait for the build to finish), serve after the build finishes, or write amp.js before building any of the extensions.

### LINKED ISSUE #12860: amp-bind does not work on LocalDev anymore
It ries to load the `worker` (`ww.js`) from CDN instead of localhost.

Might be related to recent isLocalDev changes. 

/to @rsimha-amp @choumx

### LINKED ISSUE #12868: Local mode not properly being set to localdev when switching between branches
Steps to repro:

1. Checkout a local feature branch
2. gulp clean
3. gulp build
4. gulp
5. Open localhost:8000/examples/doubleclick-rtc.amp.html, open console and verify that amp-ad-network-doubleclick-impl.js was properly loaded from the local build.
6. Kill gulp
7. Checkout a different branch
8. Run gulp, wait to allow recompilation
9. Refresh example page
10. Should now see error message in console: `extensions-impl.js:610 GET https://cdn.ampproject.org/rtv/011516204030374/v0/amp-ad-network-doubleclick-impl-0.1.js net::ERR_ABORTED`

### LINKED ISSUE #12895: `gulp` (or `gulp watch`) doesn't build everything that `gulp build` does
This is bad because we should be able to run `gulp` from a clean state. Right now, things like service worker and web worker are not built unless you run `gulp build` first.

### PR DESCRIPTION
**Pull request by rsimha-amp**
It turns out that `gulp watch` and `gulp build` had diverged over time in the build steps they would run. This was due to new build steps that were added to `build()` but not to `watch()` in PRs like #7436 ([here](https://github.com/ampproject/amphtml/pull/7436/files#diff-b9e12334e9eafd8341a6107dd98510c9R478)) and #4037 ([here](https://github.com/ampproject/amphtml/pull/4037/files#diff-b9e12334e9eafd8341a6107dd98510c9R318)).

This PR consolidates the build steps in `watch` and `build`. Now, there is no need to run `gulp build` before `gulp`. Running `gulp` will do a full build, and then wait for files to be changed.

Fixes #12895 
Fixes #12860 
Fixes #12747 
Fixes #12868
Fixes #12852

### CODE REVIEW

**Comment by rsimha-amp** [comment]
/to @cramforce @erwinmombay @choumx @jridgewell @aghassemi @lannka

(Lots of people, since this change affects the default `gulp` behavior)

File: gulpfile.js [hunk 1]
```diff
@@ -608,21 +601,24 @@ function enableLocalTesting(targetFile) {
...
+ * Performs the build steps for gulp build and gulp watch
+ * @param {boolean} watch
  * @return {!Promise}
  */
-function build() {
-  process.env.NODE_ENV = 'development';
-  printConfigHelp('gulp build');
-  return compileCss().then(() => {
+function performBuild(watch) {
+  if (!watch) {
```
Comment (cramforce) [inline gulpfile.js]: Does anyone remember why this is?

Comment (rsimha-amp) [reply gulpfile.js]: Perhaps this too was an artifact of someone enabling it for `build()` but not `watch()`? Investigating...

Comment (rsimha-amp) [reply gulpfile.js]: Yes, it was added in #678 to `build()`, but not to `watch()`. I think it's safe to do this in either case.

Comment (rsimha-amp) [reply gulpfile.js]: Done.

**Comment by rsimha-amp** [comment]
Tested this with `gulp clean && gulp`. Things like `amp-bind`, which needed a prerequisite `gulp build` now work with just `gulp`. Also tested `gulp build` to make sure all is well.

This PR is ready for a full review.

**Comment by rsimha-amp** [comment]
Added one last commit to make sure that if there's a recompilation of the runtime or the 3p frame during `watch`, we make sure that the `localDev` setting is restored correctly.

Tested this with `gulp`, `gulp build`, and `gulp dist` to make sure the normal flow remains unchanged, and that we aren't incorrectly enabling `localDev` anywhere.

**Pull request merged**